packages/dds/tree/src/forest/cursor.ts

```diff
@@ -4,89 +4,183 @@
- * Default chunks of size 1, and "node" shape?
- */
-export interface ITreeCursor<TResult = TreeNavigationResult> {
-    /** Select the child located at the given key and index. */
-    down(key: FieldKey, index: number): TResult;
+    // ********** APIs for when mode = Fields, and not pending ********** //
 
     /**
-     * Moves `offset` entries in the field.
+     * Returns the FieldKey for the current field.
+     *
+     * Allowed when `mode` is `Fields`, and not `pending`.
      */
-    seek(offset: number): TResult;
+    getCurrentFieldKey(): FieldKey;
 
-    /** Select the parent of the currently selected node. */
-    up(): TResult;
+    /**
+     * @returns the number of immediate children in the current field.
+     *
+     * Allowed when `mode` is `Fields`, and not `pending`.
+     */
+    getCurrentFieldLength(): number;
 
-    /** The type of the currently selected node. */
-    readonly type: TreeType;
+    /**
+     * Sets current node to the on at the provided `index` of the current field.
+     *
+     * Allowed when `mode` is `Fields`, and not `pending`.
+     * Sets mode to `Nodes`.
+     */
+    enterChildNode(index: number): void;
+
+    // ********** APIs for when mode = Nodes ********** //
 
     /**
-     * @returns the keys of the currently selected node.
-     * TODO: ordering invariants: Consistent over time? Consistent across nodes? Sorted?
-     * TODO: empty fields: are they always omitted here? Sometimes omitted? Depends on field kind and schema?
-     * */
-    keys: Iterable<FieldKey>;
+     * @returns a path to the current node.
+     *
+     * Only valid when `mode` is `Nodes`.
+     */
+    getPath(): UpPath;
 
-    /** @returns the number of immediate children for the given key of the currently selected node. */
-    length(key: FieldKey): number;
+    /**
+     * Index (with its parent field) of the current node.
+     *
+     * Only valid when `mode` is `Nodes`.
+     */
+    readonly currentIndexInField: number;
 
-    /** value associated with the currently selected node. */
+    /**
+     * Index (with its parent field) of the first node in the current chunk.
+     * Always less than or equal to `currentIndexInField`.
+     *
+     * Only valid when `mode` is `Nodes`.
+     */
+    readonly currentChunkStart: number;
+
+    /**
+     * Length of current chunk.
+     * Since an entire chunk always has the same `pending` value,
+     * can be used to help skip over all of a pending chunk at once.
+     *
+     * TODO:
+     * Add optional APIs to access underlying chunks so readers can
+     * accelerate processing of chunk formats they understand.
+     *
+     * Only valid when `mode` is `Nodes`.
+     */
+    readonly currentChunkLength: number;
+
+    // ********** APIs for when mode = Nodes and not pending ********** //
+
+    /**
+     * Sets the current field to the specified one and set the mode to `Fields`
```

Comment:
```suggestion
     * Navigates to the field with the specified `key` and set the mode to `Fields`
```
"Sets the current field" sounds too much like editing.

Reply:
Verbs should use the same tense:

```suggestion
     * Navigate to the field with the specified `key` and set the mode to `Fields`.
```